### PR TITLE
remove `ZERO_ADDRESS` and `is_zero`, use `Option<AccountId>` instead

### PR DESCRIPTION
**Pull request by coreggon11**
Addresses [this issue](https://github.com/Brushfam/openbrush-contracts/issues/42).

TODO:
- [x] Remove `ZERO_ADDRESS` and `is_zero()`
- [x] Use `Option<AccountId>` and `is_none()` where possible
- [x] Fix unit tests
- [ ] In many places we were checking if account is zero (for example in transfers). We can either change these functions to use `Option<AccountId>` and check if `is_none()` or leave it as it is, removing this check. This makes sense IMO, since sending tokens to `None` does not really make sense (need to think about this and either leave it or change :))

### CODE REVIEW

File: contracts/src/access/ownable/mod.rs [hunk 1]
```diff
@@ -61,33 +50,30 @@ where
     F: FnOnce(&mut T) -> Result<R, E>,
     E: From<OwnableError>,
 {
-    if instance.data().owner != T::env().caller() {
-        return Err(From::from(OwnableError::CallerIsNotOwner))
+    match instance.data().owner {
+        Some(account) if account == T::env().caller() => body(instance),
```
Comment (Artemka374) [inline contracts/src/access/ownable/mod.rs:54]: may we change the syntax here to more readable?

File: contracts/src/traits/errors/psp22.rs [hunk 2]
```diff
@@ -105,6 +105,8 @@ pub enum PSP22TokenTimelockError {
     NoTokensToRelease,
     /// Returned if the timestamp provided is before the current time
     ReleaseTimeIsBeforeCurrentTime,
+    /// Returned if either the token or the beneficiary is `None`
+    ZeroAddress,
```
Comment (Artemka374) [inline contracts/src/traits/errors/psp22.rs:109]: Let's call it another name, e.g. AccountDoesNotExist

File: contracts/src/traits/errors/timelock_controller.rs [hunk 3]
```diff
@@ -40,6 +40,7 @@ pub enum TimelockControllerError {
     MissingDependency,
     UnderlyingTransactionReverted,
     CallerMustBeTimeLock,
+    ZeroAddress,
```
Comment (Artemka374) [inline contracts/src/traits/errors/timelock_controller.rs:43]: Same here

File: contracts/src/traits/errors/psp22.rs (hunk 2 above)
Comment (Artemka374) [reply contracts/src/traits/errors/psp22.rs:109]: because this way, user can use zero address, so this naming of error looks strange

**Review by Artemka374 — CHANGES_REQUESTED** [review]
lgtm, but take a look at comments pls

**Comment by deuszx** [comment]
Pardon my ignorance, but wouldn't it be simpler to change the definition of `ZERO_ADDRESS`? I.e. make it something for which there is no known private key?

**Comment by coreggon11** [comment]
> Pardon my ignorance, but wouldn't it be simpler to change the definition of `ZERO_ADDRESS`? I.e. make it something for which there is no private key?

it would indeed, but it will be nicer to utilize `Option`

**Comment by deuszx** [comment]
> > Pardon my ignorance, but wouldn't it be simpler to change the definition of `ZERO_ADDRESS`? I.e. make it something for which there is no private key?
> 
> it would indeed, but it will be nicer to utilize `Option`

These changes here will require every downstream consumer of the lib to heavily update their contracts though while changing the `ZERO_ADDRESS` address would require no work and still mean the same thing semantically.

**Comment by coreggon11** [comment]
> These changes here will require every downstream consumer of the lib to heavily update their contracts though while changing the `ZERO_ADDRESS` address would require no work and still mean the same thing semantically.

You can still use the previous versions in your contracts and not use the zero address. The updates in this PR and in a new version will be more rusty and clear (for example, `[1u8; 32]` as `ZERO_ADDRESS ` does not really make sense).

However, a good idea is simply changing `ZERO_ADDRESS ` to, i.e., `[1u8; 32]` in the 3.1.1 patch, not letting projects using this fall for the known private key, and release the more rusty and comprehensible updates in a 3.2.0 version.

**Comment by deuszx** [comment]
There's probably already a lot of contracts that are simply Ethereum's translated to ink and b/c of that they're dependent on the fact that there exists something like a `ZERO_ADDRESS` (take for example Uniswap contract). The proposed change will require major overhaul of these contracts (which increases the likelihood of bugs).

**Pull request closed without merging**